### ISSUE TITLE
InsertAsync ignores explicit DbConnection

### ISSUE THREAD

**Issue by jeffraska** [MightyOrm/Mighty#36]
**Version of Mighty**
Mighty 3.1.3 (netstandard2.1)

**Database**
MariaDB 10.1.37

**ADO.NET Driver and version**
MySql.Data 8.0.23

**.NET version**
.NET 5.0

**What happened**
Example code:
```
private readonly DbConnection _connection;
public LicensesController(DbConnection Connection)
{
	// Initialize connection using dependency injection
	this._connection = Connection;
}
public async Task<IActionResult> AddNote(ulong Id, string Note)
{
	await using DbTransaction t = await _connection.BeginTransactionAsync();
	...
	await new MightyOrm<LicenseNote>().InsertAsync(_connection, new object[]{new LicenseNote
	{
		LicenseId = Id,
		Timestamp = DateTimeOffset.UtcNow,
		Content = Note
	}});
	...
	await t.CommitAsync();
}
```

`_connection` object is properly initialized and connection is open, but when `InsertAsync` is called, it throws exception:

```
MySql.Data.MySqlClient.MySqlException (0x80004005): Unable to connect to any of the specified MySQL hosts.
 ---> System.ArgumentException: The parameter 'host' cannot be an empty string.
   at System.Net.DnsEndPoint..ctor(String host, Int32 port, AddressFamily addressFamily)
   at System.Net.DnsEndPoint..ctor(String host, Int32 port)
   at System.Net.Sockets.Socket.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.ConnectAsync(String host, Int32 port)
   at System.Net.Sockets.TcpClient.ConnectAsync(String host, Int32 port)
   at MySql.Data.Common.StreamCreator.GetTcpStream(MySqlConnectionStringBuilder settings)
   at MySql.Data.Common.StreamCreator.GetStream(MySqlConnectionStringBuilder settings)
   at MySql.Data.MySqlClient.NativeDriver.Open()
   at MySql.Data.MySqlClient.NativeDriver.Open()
   at MySql.Data.MySqlClient.Driver.Open()
   at MySql.Data.MySqlClient.Driver.Create(MySqlConnectionStringBuilder settings)
   at MySql.Data.MySqlClient.MySqlPool.CreateNewPooledConnection()
   at MySql.Data.MySqlClient.MySqlPool.GetPooledConnection()
   at MySql.Data.MySqlClient.MySqlPool.TryToGetDriver()
   at MySql.Data.MySqlClient.MySqlPool.GetConnection()
   at MySql.Data.MySqlClient.MySqlConnection.Open()
   at System.Data.Common.DbConnection.OpenAsync(CancellationToken cancellationToken)
--- End of stack trace from previous location ---
   at Mighty.MightyOrm`1.OpenConnectionAsync(CancellationToken cancellationToken)
   at Mighty.MightyOrm`1.ScalarAsync(DbCommand command, CancellationToken cancellationToken, DbConnection connection)
   at Mighty.MightyOrm`1.ActionOnItemAsync(OrmAction originalAction, Object item, DbConnection connection, CancellationToken cancellationToken)
   at Mighty.MightyOrm`1.ActionOnItemsWithOutputAsync(OrmAction action, DbConnection connection, IEnumerable`1 items, CancellationToken cancellationToken)
   at Mighty.MightyOrm`1.ActionOnItemsAsync(OrmAction action, DbConnection connection, IEnumerable`1 items, CancellationToken cancellationToken)
   at Mighty.MightyOrm`1.InsertAsync(DbConnection connection, Object[] items)
   at Server.Controllers.LicensesController.AddNote(UInt64 Id, string Note)
```

The `ActionOnItemAsync` methods pass my `DbConnection` to `DbCommand`, but no `DbConnection` is passed to `ScalarAsync`. Then since connection parameter of `ScalarAsync` is `null`, it tries to create new connection using empty ConnectionString.

**What should have happened**
`ScalarAsync` should use `DbCommand.Connection` if property is set otherwise use current logic. And of course if this property is already set, connection should not be disposed.

**Comment by mikebeaton** [comment]
Okay, let me have a look...

**Comment by mikebeaton** [comment]
You are right that the `DbConnection` object is not being passed through in this case, and that this needs a fix; and associated new tests.

You additionally have the issue that you appear to be creating an instance of Mighty which does not know which of the [supported database providers](https://mightyorm.github.io/Mighty/docs/supported-databases.html) it is for (which should not be being permitted!), unless you have used the Mighty `GlobalConnectionString` setting (e.g. with just a ProviderName in the connection string) to set that elsewhere? (FYI Mighty as currently designed is not able to set the database provider later than in the constructor, i.e. it does not currently detect this from any `DbConnection` passed in to individual methods; and it is not realistic to update it to do so, without a major redesign.)

Apologies for the delay in updating the code to get you a fix, but I wonder could you for now inject the connection _string_, or a class which provides the connection string? For the latter, you could look at creating an instance of Mighty's `ConnectionProvider` object.

I appreciate that what you are doing should work - and requires a fix! - but that said, passing in `DbConnection` objects is meant to be for exceptional cases in Mighty (though not least for managing transactions, if that is what you are doing?); this is because all the supported providers pool connections, so the normal use case is meant to be to pass in (or perhaps globally specify, if you wish) the connection string, and to allow the underlying provider to efficiently get you a DB connection from its pool.

**Comment by jeffraska** [comment]
Oh yes, I forgot to mention that ProviderName is set through `GlobalConnectionString`.

Sadly injecting ConnectionString is not a solution since I need to manage transactions manualy.
Anyway this issue is nothing critical for me. I wrote workaround using custom SQL query and `DbCommand`. Yes, it's not nice and clean solution, but it works now.

**Comment by Bmju** [comment]
Thank you for your faith! I'll try to get out a fix asap.

**Comment by mikebeaton** [comment]
Thank you for your faith! I'll try to get out a fix asap.

**Comment by mikebeaton** [comment]
@jeffraska -

It turns out there was a lot that needed massaging and updating, to polish up both async support and explicit `DbConnection` support in the CRUD methods: New, Insert, Update, Save, Delete. I've pretty much chased through all the code paths, helped by the large - and now a little larger! - test suite, so now getting there: https://github.com/MightyOrm/Mighty/compare/fixes  .

I do need to run the last test of making sure that the new build is binary compatible with older builds, and also, probably, merging these changes on the v3 branch into the v4 branch before releasing them on v3 - both because I need to anyway, and because the v4 branch has additional checks on the consistency of method signatures.

So if possible please do bear with me, and thanks anyway for reporting something that it turns out definitely needed attention!

**Comment by MikeBeaton** [comment]
@jeffraska -

It turns out there was a lot that needed massaging and updating, to polish up both async support and explicit `DbConnection` support in the CRUD methods: New, Insert, Update, Save, Delete. I've pretty much chased through all the code paths, helped by the large - and now a little larger! - test suite, so now getting there: https://github.com/MightyOrm/Mighty/compare/fixes  .

I do need to run the last test of making sure that the new build is binary compatible with older builds, and also, probably, merging these changes on the v3 branch into the v4 branch before releasing them on v3 - both because I need to anyway, and because the v4 branch has additional checks on the consistency of method signatures.

So if possible please do bear with me, and thanks anyway for reporting something that it turns out definitely needed attention!

**Comment by mikebeaton** [comment]
I have done the binary compatibility check, so this new alpha version is already stable and usable:
https://www.nuget.org/packages/Mighty/3.2.0-alpha1

I still need to do the 'merge with 4.0.0 and check the method signatures' step; so it is possible that the order of arguments in some of the new methods added above may change slightly before release.

But the above version is safe, stable and binary and compilation compatible with 3.1.3, in case it helps you to have the new features now rather than in a few more days.

**Comment by jeffraska** [comment]
Thanks very much. I'll try to test it in my project as soon as I can.

**Comment by mikebeaton** [comment]
I have now published these updates as release [3.2.0](https://www.nuget.org/packages/Mighty/3.2.0).

The order of `DbConnection` and `CancellationToken` params will change in next major version of Mighty (was to be called v4, but will probably be called v5 now since will have explicit .NET 5 support), but that will apply to all methods, not just the few ones I newly added here - so basically what I put up as 3.2.0-alpha1 was already fine for 3.2.0.

I believe the problems you had are now fixed and tested as fixed - but do let me know if anything is not working for you.
